tests/test_py.py

```diff
@@ -2750,3 +2751,61 @@ def test_apply_not_a_function():
 # TODO: test py -i 'code ...'
 # TODO: test 'py -i' == 'py' (no double shell)
 # TODO: exiting debugger with EOF (control-D)
+
+def test_virtualenv_recognized(tmpdir, monkeypatch):
+    """Verify that virtualenv sys.path is set correctly, and that warnings are emitted."""
+    if os.environ.get("VIRTUAL_ENV") is not None:
+        old_path = os.environ["PATH"].split(os.pathsep)
+        new_path = os.pathsep.join(old_path[1:])
+
+        monkeypatch.delenv("VIRTUAL_ENV")
+        monkeypatch.setenv("PATH", new_path)
+
+    no_venv_stdout = py('print(sys.path)')
+
+    env_dir = os.path.join(tmpdir, "venv")
+    env_bin = os.path.join(env_dir, "Scripts" if os.name == "nt" else "bin")
+    venv.create(env_dir)
+
+    # Simulate activation
+    monkeypatch.setenv("VIRTUAL_ENV", env_dir)
+    monkeypatch.setenv("PATH", env_bin + os.pathsep + os.environ["PATH"])
+
+    venv_stdout = py('print(sys.path)')
+
+    # Check that the appropriate warning is in place when using the venv,
+    # and missing if not.
+    warning =  (
+        "UserWarning: Attempting to work in a virtualenv. "
+        "If you encounter problems, please install IPython inside the virtualenv."
+    )
+    assert warning not in no_venv_stdout
+    assert warning in venv_stdout
+
+    # Check that sys.path printed from the subprocess contains the same
+    # paths as what we have in the test process
+    for path in sys.path:
+
+        # If a path if missing from one, it must be missing from the other
+        # (because both are called in subprocesses, which means that e.g.
+        # the pyenv bin path won't be included in the subprocess call but
+        # will be in the pytest call that runs this test)
+        if path not in no_venv_stdout:
+            assert path not in venv_stdout
+        else:
+            assert path in venv_stdout
+            assert path in no_venv_stdout
+
+    # Check that sys.path of the non-virtualenv appears
+    # in the sys.path of the virtualenv
+    #
+    # Get the last line (which contains the printed sys.path); remove the
+    # square brackets at beginning and end which are printed because
+    # sys.path is a list
+    no_venv_stdout_path = no_venv_stdout.split('\n')[-1][1:-1]
+    assert no_venv_stdout_path in venv_stdout.split('\n')[-1]
```

Comment:
Maybe use `ast.literal_eval()` to convert back to a `list`.

Reply:
Yep, good idea.